kore/src/Kore/Step/EquationalStep.hs

```diff
@@ -353,22 +364,93 @@ See also: 'applyRewriteRule'
 applyRulesSequence
     ::  forall unifier variable
     .   ( SimplifierVariable variable
-        , Log.WithLog Log.LogMessage unifier
         , MonadUnify unifier
         )
-    => UnificationProcedure
-    -> SideCondition (Target variable)
+    => SideCondition (Target variable)
     -> Pattern (Target variable)
     -- ^ Configuration being rewritten
     -> [EqualityPattern variable]
     -- ^ Rewrite rules
     -> unifier (Results EqualityPattern variable)
 applyRulesSequence
-    unificationProcedure
     sideCondition
     initial
     -- Wrap the rules so that unification prefers to substitute
     -- axiom variables.
     (map EqualityPattern.toAxiomVariables -> rules)
-  = unifyRules unificationProcedure sideCondition initial rules
+  = unifyRules sideCondition initial rules
     >>= finalizeRulesSequence sideCondition initial
+
+-- |Unifies/matches a list a rules against a configuration. See 'unifyRule'.
+unifyRules
+    :: SimplifierVariable variable
+    => MonadUnify unifier
+    => UnifyingRule rule
+    => SideCondition (Target variable)
+    -> Pattern (Target variable)
+    -- ^ Initial configuration
+    -> [rule (Target variable)]
+    -- ^ Rule
+    -> unifier [UnifiedRule (Target variable) (rule (Target variable))]
+unifyRules sideCondition initial rules =
+    Monad.Unify.gather $ do
+        rule <- Monad.Unify.scatter rules
+        unifyRule sideCondition initial rule
+
+{- | Attempt to unify a rule with the initial configuration.
+
+The rule variables are renamed to avoid collision with the configuration. The
+rule's 'RulePattern.requires' clause is combined with the unification
+solution. The combined condition is simplified and checked for
+satisfiability.
+
+If any of these steps produces an error, then @unifyRule@ returns that error.
+
+@unifyRule@ returns the renamed rule wrapped with the combined conditions on
+unification. The substitution is not applied to the renamed rule.
+
+ -}
+unifyRule
+    :: SimplifierVariable variable
+    => MonadUnify unifier
+    => UnifyingRule rule
+    => SideCondition variable
+    -- ^ Top level condition.
+    -> Pattern variable
+    -- ^ Initial configuration
+    -> rule variable
+    -- ^ Rule
+    -> unifier (UnifiedRule variable (rule variable))
+unifyRule sideCondition initial rule = do
+    let (initialTerm, initialCondition) = Pattern.splitTerm initial
+        mergedSideCondition =
+            sideCondition `SideCondition.andCondition` initialCondition
```

Comment:
Not relevant for this PR, but we can probably remove this merge and get a performance improvement. Or make a special case for `andCondition` with top in which we don't update the representation.